### PR TITLE
add ghp-updater.sh script for generate doc pages to gh-pages

### PR DESCRIPTION
**Pull request by shiywang**
**1. about our index.html**
I take (project rust-random) https://github.com/rust-random/rand/blob/master/.travis.yml#L158-L161 and
https://github.com/huonw/travis-cargo/blob/master/travis_cargo.py#L145 
as a reference for generate our `index.html` 

`maybe we could use docstrings in the top-level crate for this?` since I do not quite understand the alternative way you suggested @steveeJ 

**2. about using travis or prow to push to gh-pages branch** 
if I added `ghp-updater.sh` to origin/release prow job, Does that ci-boot has permission to push to gh-pages by default ?

### CODE REVIEW

**Review by steveej — COMMENTED** [review]
Are the other docs reachable by the cincinnati documentation which is the landing page now?

Also some bikesheddy suggestions.

**Comment by shiywang** [comment]
@steveeJ comments addressed, the left question will be where do we get `GITHUB_TOKEN`?

if we using Travis it will be ideally we have an independent bot account which has write permission of project. 


if I added this script into prow origin/release, does it has the permission to do push to gh-pages by default ? cc @stevekuznetsov

**Comment by stevekuznetsov** [comment]
What permissions are needed? Do you need to authenticate with GitHub to push?

**Comment by steveej** [comment]
> if we using Travis it will be ideally we have an independent bot account which has write permission of project.

I think we can use that bot account on the existing CI too.
@stevekuznetsov are the GH credentials to pushing to this repo available in the regular test template or do we have to use a custom job again?

**Comment by stevekuznetsov** [comment]
I think you should turn on the auto-update feature in GitHub for the pages instead of doing this. I messaged @shiywang about this.

**Comment by steveej** [comment]
@stevekuznetsov 

> I think you should turn on the auto-update feature in GitHub for the pages instead of doing this. I messaged @shiywang about this.

Unless I'm missing something that's not going to work for us. We aren't going to use jekyll nor are we going to upload the docs in html format. We'd like the CI to render the code docstrings.

**Comment by stevekuznetsov** [comment]
>  nor are we going to upload the docs in html format.

What are you serving from the website if not HTML?

**Comment by steveej** [comment]
> What are you serving from the website if not HTML?

Sorry that was a bit misleading. We want to serve HTML but we don't want the developers to render and upload it. We rather want the CI to do that step on every merge to master.

**Comment by stevekuznetsov** [comment]
I think checking it in and using GitHub's automation to update your pages would be an order of magnitude simpler than what we'd have to string together to get this to happen.

**Comment by steveej** [comment]
> I think checking it in and using GitHub's automation to update your pages would be an order of magnitude simpler than what we'd have to string together to get this to happen.

I've done this before in Travis and it wasn't a huge deal at all. I acknowledge your warning, however I will try to integrate this with CI anyway. After all we already have a CI job which supports external secrets, so we might as well make use of it :-)

File: dist/ghp-updater.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+set -ex
+
+ABSOLUTE_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${ABSOLUTE_PATH}/commons.sh"
+TARGET_DIR="${PROJECT_PARENT_DIR}target"
+GH_REPO="shiywang/cincinnati"
```
Comment (shiywang) [inline dist/ghp-updater.sh]: will change to openshift, once postsubmit job is done

File: dist/ghp-updater.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+set -ex
+
+ABSOLUTE_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${ABSOLUTE_PATH}/commons.sh"
+TARGET_DIR="${PROJECT_PARENT_DIR}target"
```
Comment (steveej) [inline dist/ghp-updater.sh]: Is there a `/` missing or is that included in `PROJECT_PARENT_DIR`? If the latter is true we should probably remove it from there.

File: dist/ghp-updater.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+set -ex
```
Comment (steveej) [inline dist/ghp-updater.sh]: This will expose the token in the logs due to the `set -x` earlier. We should probably remove the `-x` from it:

```suggestion
set -e
```

File: dist/ghp-updater.sh (hunk 2 above)
Comment (shiywang) [reply dist/ghp-updater.sh]: `+ RELEASE_OUTPUT_DIR=/home/shiywang/ART/cincinnati/dist/..//release-20190621.104710`

Yeah, I will remove it

File: dist/ghp_updater.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+set -e
+
+ABSOLUTE_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${ABSOLUTE_PATH}/commons.sh"
+DOC_OUTPUT_DIR="${PROJECT_PARENT_DIR}/target/doc"
+GH_REPO="shiywang/cincinnati"
+FULL_REPO="https://${CINCINNATI_DOC_PUSH_TOKEN}@github.com/${GH_REPO}.git"
```
Comment (steveej) [inline dist/ghp_updater.sh]:
```suggestion
FULL_REPO="https://${CINCINNATI_GITHUB_PUSH_TOKEN}@github.com/${GH_REPO}.git"
```

File: dist/ghp_updater.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+
+# clean up target/doc before generate docs
+rm -rf ${DOC_OUTPUT_DIR}
+mkdir ${DOC_OUTPUT_DIR}
+cd ${DOC_OUTPUT_DIR}
+
+if [[ -z "${CINCINNATI_DOC_PUSH_TOKEN}" ]]; then
```
Comment (steveej) [inline dist/ghp_updater.sh]:
```suggestion
if [[ -z "${CINCINNATI_GITHUB_PUSH_TOKEN}" ]]; then
```

File: dist/ghp_updater.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+	echo 'No CINCINNATI_DOC_PUSH_TOKEN ENV found, do not push to the remote repo'
```
Comment (steveej) [inline dist/ghp_updater.sh]:
```suggestion
	echo 'No CINCINNATI_GITHUB_PUSH_TOKEN ENV found, do not push to the remote repo'
```

File: dist/ghp_updater.sh [hunk 7]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+set -e
+
+ABSOLUTE_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${ABSOLUTE_PATH}/commons.sh"
+DOC_OUTPUT_DIR="${PROJECT_PARENT_DIR}/target/doc"
+GH_REPO="shiywang/cincinnati"
```
Comment (steveej) [inline dist/ghp_updater.sh:8]: We should respect `CINCINNATI_GITHUB_PUSH_TOKEN_PATH` which could exist in CI and point to a file rather than containing the token directly.

```suggestion
GH_REPO="shiywang/cincinnati"
CINCINNATI_GITHUB_PUSH_TOKEN="${CINCINNATI_GITHUB_PUSH_TOKEN:-$(cat ${CINCINNATI_GITHUB_PUSH_TOKEN_PATH})"
```

File: dist/ghp_updater.sh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+if [[ -z "${CINCINNATI_DOC_PUSH_TOKEN}" ]]; then
+	echo 'No CINCINNATI_DOC_PUSH_TOKEN ENV found, do not push to the remote repo'
+else
+	git init
+	git remote add doc "${FULL_REPO}"
+	git config user.name "openshift-merge-robot"
+	git config user.email "openshift-merge-robot@redhat.com"
+	git fetch doc gh-pages
+    git checkout gh-pages
```
Comment (steveej) [inline dist/ghp_updater.sh]: inconsistent line indentation here

File: dist/ghp_updater.sh [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    cargo doc --no-deps --all --all-features
```
Comment (steveej) [inline dist/ghp_updater.sh]: inconsistent line indentation here

File: dist/ghp_updater.sh [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    echo '<meta http-equiv=refresh content=0;url=cincinnati/index.html>' > index.html
```
Comment (steveej) [inline dist/ghp_updater.sh]: inconsistent line indentation here

File: dist/ghp_updater.sh [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+	# commit and push changes
+    git add -A
```
Comment (steveej) [inline dist/ghp_updater.sh]: inconsistent line indentation here

**Review by steveej — COMMENTED** [review]
I adjusted the file to my review on https://github.com/openshift/release/pull/4138.
I also pointed out the inconsistent line indentations. I don't have a strong opinion on what indentation we use, just that it doesn't change within the same script ;-)

File: dist/ghp_updater.sh [hunk 12]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+set -e
+
+ABSOLUTE_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${ABSOLUTE_PATH}/commons.sh"
+DOC_OUTPUT_DIR="${PROJECT_PARENT_DIR}/docs"
+GH_REPO="shiywang/cincinnati"
```
Comment (steveej) [inline dist/ghp_updater.sh:8]:
```suggestion
GH_REPO="${GH_REPO:-openshift/cincinnati}"
```

File: dist/ghp_updater.sh [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+FULL_REPO="git@github.com:${GH_REPO}.git"
+
+
+if [[ -z "${CINCINNATI_GITHUB_PUSH_TOKEN_PATH}" ]]; then
+	echo 'No CINCINNATI_GITHUB_PUSH_TOKEN_PATH found, do not push to the remote repo'
+else
+	ssh-add ${CINCINNATI_GITHUB_PUSH_TOKEN_PATH}
+	cargo doc --no-deps --all --all-features --target-dir=${DOC_OUTPUT_DIR} --release
+	echo '<meta http-equiv=refresh content=0;url=doc/cincinnati/index.html>' > ${DOC_OUTPUT_DIR}/index.html
```
Comment (steveej) [inline dist/ghp_updater.sh:17]: I don't think we need to write this file every time. Can you please check this file in manually and remove this line?

File: dist/ghp_updater.sh [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+	# commit and push changes
+	git add -A
+	git commit -m "GH-Pages update by openshift-merge-robot"
+	git push origin master
```
Comment (steveej) [inline dist/ghp_updater.sh:21]: I would prefer if we used the `gh-pages` branch for the documentation like in the previous iterations of this script. Why did you change the approach?

Comment (shiywang) [reply dist/ghp_updater.sh:21]: In postsubmit job you mentioned we are using master branch instead of gh-pages?

Comment (steveej) [reply dist/ghp_updater.sh:21]: Ah, I think that's a misunderstanding. The postsubmit job has to run on merge to master, but the job itself should (in my opinion) push the doc updates to the gh-pages branch.

Comment (shiywang) [reply dist/ghp_updater.sh:21]: ah.... I see, thank you for the clarification, I have very little time working on Cincinnati this week, I will update the PR this weekends if I got chance.
